### PR TITLE
Improve reading CSV field in CustomSeparated/Template format

### LINKED ISSUE #39620: CustomSeparated/Template formats don't work properly with CSV escaping rule and arbitrary field delimiter
版本为：22.6.3.35

问题：使用CustomSeparated 格式导入多字符字段分隔符的文本中，文本内容不能包含多分隔中的第一个字符。

测试过程：
 * 创建表
```
CREATE TABLE tableA
(
    `c1` text,
    `c2` text,
    `c3` text
)
ENGINE = MergeTree
ORDER BY c1
```
 
 * cat tableA.txt
```
name1|!name2|!name3
name41|!name42|!name43
name51|!name|52|!name53
```
文件的分隔符号为 ”|!“
当导入tableA.txt 数据的时候，就会在name|52位置报错。 但是”|“ 字符是我需要的内容

```
[root@hostname ~]# cat tableA.txt | clickhouse client --query "insert into tableA FORMAT CustomSeparated" --format_custom_row_after_delimiter=$'\x0A' --format_custom_escaping_rule='CSV' --format_custom_field_delimiter=$'\x7C\x21' --max_insert_block_size=1000000
Code: 27. DB::ParsingException: Cannot parse input: expected '|!' before: '52|!name53\n':
Row 2:
Column 0,   name: c1, type: String, parsed text: "name41"
Column 1,   name: c2, type: String, parsed text: "name42"
Column 2,   name: c3, type: String, parsed text: "name43"

Row 3:
Column 0,   name: c1, type: String, parsed text: "name51"
Column 1,   name: c2, type: String, parsed text: "name"
ERROR: There is no delimiter between fields: expected "|!", got "52|!name53<LINE FEED>"

: While executing CustomSeparatedRowInputFormat: data for INSERT was parsed from stdin: (in query: insert into tableA FORMAT CustomSeparated): (at row 3)
. (CANNOT_PARSE_INPUT_ASSERTION_FAILED)

```
就像报错提示的内容一样，如果字段中含有分隔符（多个）的首个字符，则解析会认为是分隔符开始，但是这不是我期望的结果。因为`name|52` 就是一个完整的字符串。

### LINKED ISSUE #42352: CustomSeparated is confused without quotes
Tried on CH version 22.10.1.1232

This is pretty similar to --input-format CSV, but using '|' as row delimiter instead of '\n'. Doesn't work.

```sh
echo -ne "a,b|c,d|" | clickhouse-local --structure "a String, b String"  --input-format CustomSeparated --format_custom_escaping_rule=CSV --format_custom_field_delimiter=',' --format_custom_row_after_delimiter='|' -q 'select * from table'

Code: 27. DB::Exception: Cannot parse input: expected '|' before: ',d|':
Row 1:
Column 0,   name: a, type: String, parsed text: "a"
Column 1,   name: b, type: String, parsed text: "b|c"
ERROR: There is no delimiter after last field: expected "|", got ",d|"

: While executing CustomSeparatedRowInputFormat: While executing File. (CANNOT_PARSE_INPUT_ASSERTION_FAILED)
```

This is like --input-format TSV, but using ',' instead of '\t' for delimiting columns. Doesn't work.
```sh
echo -ne "a,b\nc,d\n" | clickhouse-local --structure "a String, b String"  --input-format CustomSeparated --format_custom_escaping_rule=Escaped --format_custom_field_delimiter=',' --format_custom_row_after_delimiter=$'\n' -q 'select * from table'

Code: 27. DB::Exception: Cannot parse input: expected ',' before: '\nc,d\n':
Row 1:
Column 0,   name: a, type: String, parsed text: "a,b"
ERROR: There is no delimiter between fields: expected ",", got "<LINE FEED>c,d<LINE FEED>"

: While executing CustomSeparatedRowInputFormat: While executing File. (CANNOT_PARSE_INPUT_ASSERTION_FAILED)
```

If we fully emulate --input-format TSV using CustomSeparated, then it works without quotes. The only difference from the 2nd query is that we use original TSV - '\t' separator for fields, instead of custom one.
```sh
echo -ne "a\tb\nc\td\n" | clickhouse-local --structure "a String, b String"  --input-format CustomSeparated --format_custom_escaping_rule=Escaped --format_custom_field_delimiter=$'\t' --format_custom_row_after_delimiter=$'\n' -q 'select * from table'
a       b
c       d
```

Emulating CSV using CustomSeparated also works, no quotes needed. The only difference from the 1st query is that we are using original CSV - '\n' line separator.
```sh
echo -ne "a,b\nc,d\n" | clickhouse-local --structure "a String, b String"  --input-format CustomSeparated --format_custom_escaping_rule=CSV --format_custom_field_delimiter=',' --format_custom_row_after_delimiter=$'\n' -q 'select * from table'
a       b
c       d
```

Just --input-format CSV also works
```sh
echo -ne "a,b\nc,d\n" | clickhouse-local --structure "a String, b String"  --input-format CSV -q 'select * from table'
```

If we really want to use custom separators that are different from standard CSV or TSV, we should use --format_custom_escaping_rule=CSV and wrap strings in double quotes
```sh
echo -ne '"a","b"|"c","d"|' | clickhouse-local --structure "a String, b String"  --input-format CustomSeparated --format_custom_escaping_rule=CSV --format_custom_field_delimiter=',' --format_custom_row_after_delimiter='|' -q 'select * from table'
a       b
c       d
```

Or use single quotes with --format_custom_escaping_rule=Quoted 
```sh
echo -ne "'a','b'|'c','d'|" | clickhouse-local --structure "a String, b String"  --input-format CustomSeparated --format_custom_escaping_rule=Quoted --format_custom_field_delimiter=',' --format_custom_row_after_delimiter='|' -q 'select * from table'
```

Maybe I'm missing something, but I would expect first 2 commands to work without adding quotes, as there are just simple strings without spaces or fancy characters.

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve reading CSV field in CustomSeparated/Template format.
Closes https://github.com/ClickHouse/ClickHouse/issues/42352
Closes https://github.com/ClickHouse/ClickHouse/issues/39620

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/HDFS/StorageHDFS.cpp [hunk 1]
```diff
@@ -461,6 +461,7 @@ class HDFSSink : public SinkToStorage
     {
         std::lock_guard lock(cancel_mutex);
         finalize();
+        finalize();
```
Comment (antonio2368) [inline src/Storages/HDFS/StorageHDFS.cpp:464]: Why?

File: src/Storages/HDFS/StorageHDFS.cpp [hunk 2]
```diff
@@ -474,6 +475,7 @@ class HDFSSink : public SinkToStorage
...
```
Comment (antonio2368) [inline src/Storages/HDFS/StorageHDFS.cpp:478]: Why?

File: src/Storages/HDFS/StorageHDFS.cpp (hunk 1 above)
Comment (Avogar) [reply src/Storages/HDFS/StorageHDFS.cpp:464]: It was added while testing, it's not needed, just forgot to remove. Thanks for noticing!

Comment (Avogar) [inline src/Storages/HDFS/StorageHDFS.cpp:464]:
```suggestion
```

File: src/Storages/HDFS/StorageHDFS.cpp [hunk 3]
```diff
@@ -474,6 +474,7 @@ class HDFSSink : public SinkToStorage
...
```
Comment (Avogar) [inline src/Storages/HDFS/StorageHDFS.cpp:477]:
```suggestion
```

**Review by antonio2368 — APPROVED** [review]

**Comment by Avogar** [comment]
`AST fuzzer (asan)` - https://github.com/ClickHouse/ClickHouse/issues/43204

**Pull request merged**